src/libcephfs.cc

```diff
@@ -2019,31 +2019,118 @@ extern "C" int64_t ceph_ll_writev(class ceph_mount_info *cmount,
   return (cmount->get_client()->ll_writev(fh, iov, iovcnt, off));
 }
 
+struct ceph_ll_readv_writev_buffer {
+  ceph_ll_readv_writev_buffer() : iov() {};
+  ~ceph_ll_readv_writev_buffer() {
+    delete iov;
+  }
+  void prepare_iovs() {
+    ceph_assert(bl.get_num_buffers() <= IOV_MAX);
+
+    iovcnt = bl.get_num_buffers();
+    iov = new struct iovec [iovcnt];
+    auto pb = std::cbegin(bl.buffers());
+    for (auto n = 0; n < iovcnt; n++) {
+      iov[n].iov_base = (void*)(pb->c_str());
+      iov[n].iov_len = pb->length();
+      ++pb;
+    }
+  }
+  bufferlist bl;
+  int iovcnt;
+  struct iovec *iov;
+};
+
+extern "C"  void LL_onfinish_release(void *);
+
 class LL_Onfinish : public Context {
 public:
   LL_Onfinish(struct ceph_ll_io_info *io_info)
-    : io_info(io_info) {}
-  bufferlist bl;
+    : buf(), io_info(io_info) {}
+  struct ceph_ll_readv_writev_buffer buf;
 private:
   struct ceph_ll_io_info *io_info;
+  void complete(int r) override {
+    finish(r);
+    if (io_info->write || io_info->zerocopy) {
+      // this is a write or a non-zerocopy read, delete...
```

Comment:
The condition finishes the context either - it is a normal write, either zerocopy write or it is a zerocopy read but for a normal read(non-zerocopy read) this will be skipped?

Reply:
Added a comment:

// for a zero-copy read, we need to keep this object until the caller
// is done with the buffers so it can use this as a hook to release them.